File: backend/app/api/deps.py
"""
API Dependencies Module

Provides reusable dependency injection functions for FastAPI endpoints.

Key Dependencies:
    - get_db: Database session management with automatic cleanup
    - get_current_user: JWT token validation and user retrieval
    - reusable_oauth2: OAuth2 password bearer token scheme

Authentication Flow:
    Request → OAuth2PasswordBearer → JWT decode → User lookup → Endpoint

Database Session Lifecycle:
    Endpoint call → SessionLocal() → Yield to endpoint → db.close()

Technical Notes:
    - OAuth2 tokenUrl points to /api/v1/auth/login
    - JWT verification uses HS256 algorithm
    - Sessions auto-closed even on exceptions (finally block)
    - 403 Forbidden for invalid tokens
    - 404 Not Found for non-existent users
"""

from typing import Generator, Optional
from fastapi import Depends, HTTPException, status
from fastapi.security import OAuth2PasswordBearer
from jose import jwt, JWTError
from pydantic import ValidationError
from sqlalchemy.orm import Session

from app.core import security
from app.core.config import settings
from app.core.database import SessionLocal
from app.models import user as models
from app.schemas import user as schemas

reusable_oauth2 = OAuth2PasswordBearer(
    tokenUrl=f"{settings.API_V1_STR}/auth/login"
)

def get_db() -> Generator:
    """
    Provide database session with automatic cleanup.
    
    Creates SQLAlchemy session for endpoint use and ensures
    proper closure even on exceptions.
    
    Yields:
        Session: SQLAlchemy database session
        
    Technical Notes:
        - Session created from SessionLocal factory
        - Automatically closed in finally block
        - Thread-safe session management
        - Used as FastAPI dependency: db: Session = Depends(get_db)
    """
    try:
        db = SessionLocal()
        yield db
    finally:
        db.close()

def get_current_user(
    db: Session = Depends(get_db), token: str = Depends(reusable_oauth2)
) -> models.User:
    """
    Validate JWT token and retrieve authenticated user.
    
    Decodes JWT token from Authorization header, validates signature
    and expiration, then retrieves user from database.
    
    Args:
        db (Session): Database session from get_db dependency
        token (str): JWT token from OAuth2PasswordBearer (Authorization header)
        
    Returns:
        User: Authenticated user object from database
        
    Raises:
        HTTPException 403: Invalid token signature, expired token, or validation error
        HTTPException 404: Token valid but user not found in database
        
    Example:
        @router.get("/me")
        def read_current_user(current_user: User = Depends(get_current_user)):
            return current_user
            
    Technical Notes:
        - Verifies token with SECRET_KEY and ALGORITHM (HS256)
        - Extracts user_id from token 'sub' claim
        - Queries database for user with extracted ID
        - Used as dependency in protected endpoints
        - Token must be in format: "Bearer <token>"
    """
    try:
        payload = jwt.decode(
            token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM]
        )
        token_data = schemas.TokenPayload(**payload)
    except (JWTError, ValidationError):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Could not validate credentials",
        )
    user = db.query(models.User).filter(models.User.id == token_data.sub).first()
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
    return user

File: backend/src/api/deps.ts
/**
 * Dependency Injection Module
 *
 * Provides reusable dependencies for route handlers:
 * - Database session
 * - Current authenticated user
 *
 * Equivalent to Python's deps.py with get_db and get_current_user
 */

import { Request, Response, NextFunction } from "express";
import { verifyToken } from "../core/security";
import { User } from "../models/user";

/**
 * Extended Express Request interface with user property
 */
export interface AuthRequest extends Request {
  user?: User;
}

/**
 * Middleware to extract and verify JWT token
 * Attaches user object to request if token is valid
 *
 * @throws 401 if token is missing or invalid
 * @throws 403 if user is inactive
 */
export async function getCurrentUser(
  req: AuthRequest,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    // Extract token from Authorization header
    const authHeader = req.headers.authorization;
    if (!authHeader || !authHeader.startsWith("Bearer ")) {
      res.status(401).json({ detail: "Not authenticated" });
      return;
    }

    const token = authHeader.substring(7); // Remove 'Bearer ' prefix

    // Verify token
    const payload = verifyToken(token);
    if (!payload) {
      res.status(401).json({ detail: "Invalid token" });
      return;
    }

    // Fetch user from database
    const user = await User.findByPk(payload.sub);
    if (!user) {
      res.status(401).json({ detail: "User not found" });
      return;
    }

    // Check if user is active
    if (!user.is_active) {
      res.status(403).json({ detail: "Inactive user" });
      return;
    }

    // Attach user to request object
    req.user = user;
    next();
  } catch (error) {
    console.error("❌ Authentication error:", error);
    res.status(401).json({ detail: "Authentication failed" });
  }
}

/**
 * Middleware to check if user is admin
 * Must be used after getCurrentUser middleware
 *
 * @throws 403 if user is not admin
 */
export function requireAdmin(
  req: AuthRequest,
  res: Response,
  next: NextFunction
): void {
  if (!req.user || !req.user.is_admin) {
    res.status(403).json({ detail: "Admin privileges required" });
    return;
  }
  next();
}
